### ISSUE TITLE
False positive - SQL injection when using IDs in second query

### ISSUE THREAD

**Issue by JasonBarnabe** [presidentbeef/brakeman#1005]
```ruby
newest_sv_ids = Script.connection.select_values('SELECT MAX(id) FROM script_versions GROUP BY script_id')
script_to_code_ids = Script.connection.select_rows("SELECT script_id, script_code_id FROM script_versions WHERE ID IN (#{newest_sv_ids.join(',')})")
```
Could brakeman see based on the SQL that the values used are from script_versions.id, and then look up to see that that's a numeric field, so there's vulnerability here?

**Comment by JasonBarnabe** [comment]
To walk this back one step... This should not be a fail.

```ruby
User.connection.execute("DELETE FROM users WHERE id IN (#{[1,2,3].join(',')})")
```

(Of course in this case I can do it much better with ActiveRecord, but this is a simplified example.) It seems like there is no way to "quote" an Array to use with a DB `IN`?

**Comment by JasonBarnabe** [comment]
Using the protected `sanitize_sql_array` seems to satisfy brakeman. No warnings on these:
```ruby
User.connection.execute("DELETE FROM users WHERE #{ActiveRecord::Base.send(:sanitize_sql_array, ['id IN (?)', [1,2,3]])}")
User.connection.execute(ActiveRecord::Base.send(:sanitize_sql_array, ['DELETE FROM users WHERE id IN (?)', [1,2,3]]))
```

**Comment by presidentbeef** [comment]
Yeah, Rails really does not make it easy to safely manually build SQL.

To answer your original question, no Brakeman cannot "tell" `newest_sv_ids.join(',')` is safe unless `newest_sv_ids` happens to be a local array with integer literals in it. Brakeman could potentially ignore `[1,2,3].join(',')` but I doubt code that obvious comes up much.

**Comment by JasonBarnabe** [comment]
I agree, no one would use that with `[1,2,3]`. But a much more common scenario would be something like this:

```ruby
user_ids = User.where(active: true).map(&:id)
User.connection.execute("DELETE FROM users WHERE id IN (#{user_ids.join(',')})")
```

Or any other time the variable you're injecting comes from an ActiveRecord object and the type could theoretically be known. I'm unsure of the limits of what brakeman can "see", so I'll defer to your judgment on this.

I had also expected (hoped?) brakeman could understand something like this:

```ruby
User.connection.execute("DELETE FROM users WHERE id IN (#{user_ids.map(&:to_i).join(',')})")
```

**Comment by presidentbeef** [comment]
MMmmmmmmm yeah I suppose it could.

**Comment by lostapathy** [comment]
Why not run the query from the 2nd comment like this?

```
user_ids = [1,2,3]
User.where(:id => user_ids).delete_all
```

**Comment by shekhar-patil** [comment]
@presidentbeef Have we updated something related to this issue?

**Comment by presidentbeef** [comment]
@shekhar-patil Don't think so.
